File: src/main/java/mnm/mods/tabbychat/util/Translatable.java
package mnm.mods.tabbychat.util;

import net.minecraft.client.resources.I18n;

/**
 * Translatable strings. Translations are done via
 * {@link I18n#format(String, Object...)}.
 *
 * @author Matthew
 */
public interface Translatable {

    /**
     * Gets the unlocalized string for this translation
     *
     * @return The untranslated string
     */
    String getUnlocalized();

    /**
     * Translates this string.
     *
     * </pre>
     *
     * @param params Translation parameters
     * @return The translated string
     */
    default String translate(Object... params) {
        return I18n.format(getUnlocalized(), params);
    }
}

File: src/main/kotlin/mnm/mods/tabbychat/util/Translatable.kt
package mnm.mods.tabbychat.util

import net.minecraft.client.resources.I18n
import net.minecraft.util.text.ITextComponent
import net.minecraft.util.text.TranslationTextComponent

inline class Translatable(private val key: String) {
    fun toComponent(vararg args: Any): ITextComponent {
        return TranslationTextComponent(this.key, *args)
    }

    fun translate(vararg args: Any): String {
        return I18n.format(this.key, *args)
    }
}